kms/cloudkms/decrypter.go

```diff
@@ -0,0 +1,120 @@
+package cloudkms
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	"cloud.google.com/go/kms/apiv1/kmspb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"go.step.sm/crypto/kms/apiv1"
+	"go.step.sm/crypto/pemutil"
+)
+
+// CreateDecrypter implements the apiv1.Decrypter interface and returns
+// a crypto.Decrypter backed by a decryption key in Google Cloud KMS.
+func (k *CloudKMS) CreateDecrypter(req *apiv1.CreateDecrypterRequest) (crypto.Decrypter, error) {
+	if req.DecryptionKey == "" {
+		return nil, errors.New("decryption key cannot be empty")
+	}
+	return NewDecrypter(k.client, req.DecryptionKey)
+}
+
+// Decrypter implements a crypto.Decrypter using Google Cloud KMS.
+type Decrypter struct {
+	client        KeyManagementClient
+	decryptionKey string
+	publicKey     crypto.PublicKey
+}
+
+// NewDecrypter creates a new crypto.Decrypter backed by the given
+// Google Cloud KMS decryption key.
+func NewDecrypter(client KeyManagementClient, decryptionKey string) (*Decrypter, error) {
+	// Make sure that the key exists.
+	decrypter := &Decrypter{
+		client:        client,
+		decryptionKey: decryptionKey,
+	}
+	if err := decrypter.preloadKey(decryptionKey); err != nil { // TODO(hs): (option for) lazy load instead?
+		return nil, err
+	}
+
+	return decrypter, nil
+}
+
+func (d *Decrypter) preloadKey(signingKey string) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+
+	response, err := d.client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{
+		Name: signingKey,
+	})
+	if err != nil {
+		return fmt.Errorf("cloudKMS GetPublicKey failed: %w", err)
+	}
+
+	d.publicKey, err = pemutil.ParseKey([]byte(response.Pem))
+	return err
+}
+
+// Public returns the public key of this decrypter
+func (d *Decrypter) Public() crypto.PublicKey {
+	return d.publicKey
+}
+
+// Decrypt decrypts ciphertext using the decryption key backed by Google Cloud KMS and returns
+// the plaintext bytes. An error is returned when decryption fails. Google Cloud KMS only supports
+// RSA keys with 2048, 3072 or 4096 bits and will always use OAEP. Labels are not supported.
+//
+// Also see https://cloud.google.com/kms/docs/algorithms#asymmetric_encryption_algorithms.
+func (d *Decrypter) Decrypt(rand io.Reader, ciphertext []byte, opts crypto.DecrypterOpts) ([]byte, error) {
+	if ropts, ok := opts.(*rsa.OAEPOptions); ok && ropts != nil {
+		if len(ropts.Label) > 0 {
+			return nil, errors.New("cloudKMS does not support RSA-OAEP label")
+		}
+		switch ropts.Hash {
+		case crypto.SHA1, crypto.SHA256, crypto.SHA512:
+			break
+		default:
+			return nil, fmt.Errorf("cloudKMS does not support hash algorithm %q with RSA-OAEP", ropts.Hash)
+		}
+	}
+	if _, ok := opts.(*rsa.PKCS1v15DecryptOptions); ok {
+		return nil, errors.New("cloudKMS does not support PKCS #1 v1.5 decryption")
+	}
```

Comment:
Let's add a typed switch here, similar to rsa.Decrypt(). The difference would be that if opts == nil the default algorithm would be RSA-OAEP instead of PKCS #1 v1.5. We should explicitly say that in the comment.

Reply:
I tried that, but the code doesn't look cleaner then. The reason for that is that there can be typed `nil`s, but also untyped `nil`s. Both have to be handled to be able to verify the input. 

The goal of the code is to show a more descriptive error message in case decryption fails if a user tries to use a parameter that we know will for sure result in an error. For example, when a label is used, the error is very generic: `Error: error decrypting input: cloudKMS AsymmetricDecrypt failed: rpc error: code = InvalidArgument desc = Decryption failed.`. Same is true if someone tries to use this with PKCS1 options. So we're not actually passing the options to GCP; we're merely checking for them, and return an error for things that GCP doesn't support.